vmm/src/lib.rs

```diff
@@ -256,24 +348,33 @@ impl Vmm {
         #[cfg(not(feature = "acpi"))]
         {
             if let Some(ref mut vm) = self.vm {
+                let mut vm = vm.lock().unwrap();
                 vm.shutdown()?;
                 return Ok(());
             }
         }
 
         // First we stop the current VM and create a new one.
         if let Some(ref mut vm) = self.vm {
+            let mut vm = vm.lock().unwrap();
             let config = vm.get_config();
             vm.shutdown()?;
 
-            let exit_evt = self.exit_evt.try_clone().map_err(VmError::EventFdClone)?;
-            let reset_evt = self.reset_evt.try_clone().map_err(VmError::EventFdClone)?;
+            self.vm_config = Some(config)
+        }
 
-            self.vm = Some(Vm::new(config, exit_evt, reset_evt)?);
+        let exit_evt = self.exit_evt.try_clone().map_err(VmError::EventFdClone)?;
+        let reset_evt = self.reset_evt.try_clone().map_err(VmError::EventFdClone)?;
+
+        if let Some(ref vm_config) = self.vm_config {
+            let vm = Vm::new(Arc::clone(vm_config), exit_evt, reset_evt)?;
+            self.vm = Some(Arc::new(Mutex::new(vm)));
+            self.migration_register().expect("Fail to register migration state component");
```

Comment:
I feel things would be cleaner if the VM, and not the VMM, would register itself.

Reply:
I mentioned the reason in cover letter. The reasons to register Vmm but not Vm are below:
1. On source side, we need get Vm's states (e.g. VmConfig) when request comes. It is hard to do this in a child thread in Vm.
2. On target side, we receive the states first and use the states to create Vm instance. It is easy to do this in Vmm (Vm may need provide new interface to create itself according to input states).

Reply:
> 1. On source side, we need get Vm's states (e.g. VmConfig) when request comes. It is hard to do this in a child thread in Vm.
>

If the Vm registers itself as a migratable component, why would it be harder to get its state compared to any other component?

> 2. On target side, we receive the states first and use the states to create Vm instance. It is easy to do this in Vmm (Vm may need provide new interface to create itself according to input states).
>

On the target side, the Vmm will hold the snapshot states. Then it creates a Vm (`vm::new()`) and the first thing a Vm does is registering itself as a migratable component. It gets its state back as a reply to the registration, and can then continue initializing.

I think registering the Vm is fine, but I may be missing your point. Please let me know if that's the case.

Reply:
> > 1. On source side, we need get Vm's states (e.g. VmConfig) when request comes. It is hard to do this in a child thread in Vm.
> 
> If the Vm registers itself as a migratable component, why would it be harder to get its state compared to any other component?
> 
If so, we need implement "migration_request_handle(&self, ...)" in Vm. A new thread will start to wait for "MigrationStateGet" command. Because we cannot get latest states before thread starts, we have to find a way to be able to clone "&self" and use it in the thread (Vm is the root to be able to access memory/devices/etc). I tried some ways but all failed.

> > 1. On target side, we receive the states first and use the states to create Vm instance. It is easy to do this in Vmm (Vm may need provide new interface to create itself according to input states).
> 
> On the target side, the Vmm will hold the snapshot states. Then it creates a Vm (`vm::new()`) and the first thing a Vm does is registering itself as a migratable component. It gets its state back as a reply to the registration, and can then continue initializing.
> 
Ok, this should be workable.

One concern is, if we implement "trait Migratable" for Vm and send it as a parameter in MigrationRegister, is that possible to do it before Vm instance is created?

> I think registering the Vm is fine, but I may be missing your point. Please let me know if that's the case.

Reply:
> One concern is, if we implement "trait Migratable" for Vm and send it as a parameter in MigrationRegister, is that possible to do it before Vm instance is created?

Ah, good point. That could be extended to any migratable component I guess.
So we're looking at 2 different steps when such component (Vm, virtio device, etc) initializes:

* Early in the initialization process, the component sends a command to fetch its state back. An empty reply means it can initialize normally, i.e. not from a migrated state
* Late in the initialization process, the component sends the actual `MigrationRegister` command but this time does not receive a payload with its state back. It always gets an empty response back.

What do you think?

Reply:
So, your proposal is to define a new ApiRequest to get states. The MigrationRegister is just for register the component. Although this way is workable, it makes things complex. Why cannot we think the upper layer be the state handler? It looks easy and more clean.

Furthermore, there is one technical issue (described above) cannot be solved. After registering, a new thread in Vm will start to wait for "MigrationStateGet" command. Because we cannot get latest states before thread starts, we have to find a way to be able to clone "&self" and use it in the thread (Vm is the root to be able to access memory/devices/etc). I tried some ways but all failed.